test/rules/no-null-undefined-union/test.ts.lint

```diff
@@ -52,4 +58,18 @@ function someFunc() {
 }
 ~ [0]
 
+const z = testFunc();
+
+type Text = string | null
+
+interface TextInterface {
+  foo?: Text
```

Comment:
1. indentation is inconsistent on this line (use 4 spaces)
2. I don't agree with this change... this is just syntactic sugar for `Text | undefined`, which is in turn sugar for `string | null | undefined`. Why should we allow this? What's the point of the rule? How do you explain to devs which style is allowed and which isn't?

Reply:
Plus `?` is just not syntactic sugar for Text | undefined.
```
interface A {
 a: number | undefined
}
interface B {
 b?: number
}

const a: A = {} // $error
const b: B = {} // no error
```

When you declare props in react, all I saw
```
interface Props {
  requiredValue: SomeType
  optionalValue?: SomeType
}

<Component requiredValue={...}  />
```

I tried
```
interface Props {
  value: SomeType | undefined
}

<Component /> // $error property value is missing
```

Plus actually the rule create conflict with the `non-null-keyword` in the `tslint:all` config. With the precedent props, you have to use the component this way
```
interface Props {
  value: SomeTypeCouldBeNull
}

<Component value={null} /> // $error prefer use undefined
```

Every user can disable the rules he does not want, but at least the proposal `tslint:all` should not have conflict.

Reply:
ok, that's fair, thanks for the examples. after looking at this again, the rule makes more sense now. I think we should update the documentation to also explain this, though. we should mention that interface fields and function params with `?` are expected to be exempt from the rule.